design/lifecycle.md

```diff
@@ -0,0 +1,195 @@
+We would basically try to remove the same addresses again and again with each run adding some new addresses.
+The main issue with running on all commits is adding unnecessary work that would get extremely large over time.
+For that reason, we would like to reduce the number of commits based on previous lifecycle runs (that finished and deleted successfully).
+
+##### Definitions
+
+- potential commits - commits that may contain addresses that should be removed as part of the lifecycle policy and weren't removed as part of the last lifecycle run.
+- commit branches - all the branches that have access to the commit
+
+#### algorithm
+
+From the `current_rule` table we extract for each branch the minimum `last_deleted`
+
+| branch | last_deleted  |
+|--------|---------------|
+|        |   09/01/1998  |
+|   b1   |               |
+|   b2   |   17/01/1998  |
+
+To enrich the example lets add another branch b3 with last_deleted 05/01/1998
+
+| branch | last_deleted  |
+|--------|---------------|
+|        |   09/01/1998  |
+|   b1   |               |
+|   b2   |   17/01/1998  |
+|   b3   |   05/01/1998  |
+
+* Empty branch == All
+* Empty last_deleted == All (the beginning of the branch)
+We pass that data to lakeFS
+
+In lakeFS we need to find all the commits matching all Dates
+That is 
+- all commits on b1
+- all commits on b2 after 17/01/1998 -> (we will take because it's part of all )
+- all commits on b3 after 05/01/1998
+- all commits on any branch after 09/01/1998
+
+We would like to take all commits from the intersection of these rules
+Some examples
+- commit is only accessible from the head of b1 - all commits
+- commit is only accessible from the head of b3 - all commits after 05/01/1998
+- commit is only accessible from the head of b1 or b3 - commits after 05/01/1998
+- commit is only accessible from the head of b1 and b3 and b2 - commits after 09/01/1998
+
+> Note! This could be improved by doing these intersections for each rule independently, It could potentially reduce the number of commits, I believe it is not worth the complication for now.   
+
+lakeFS will receive this data and return/write all the potential commits (with commit branches) for this run.
+
+#### Find addresses that should be removed by lifecycle policy
+
+From the previous stage we got a table of the following form commit_id | branches
+Using that we:
+1. get rang_id from commit_id
+2. group by range_id selecting the union of branches
+3. get (key, address) from ranges
+4. group by (keys, address) selecting the union of branches
+5. select all addresses matching the policy rules (which did not match policy rule last run)
+
+#### Find deduplications
+
+##### Definitions
+deduplications - for a given path P with address A, a deduplication would be a different path P` referencing the same address A
+
+In the case of lifecycle, before removing an address, we would first need to check that it has no deduplications that should not be removed as part of lifecycle.
```

Comment:
What about deduplications from commits that were not taken, because of the last_deleted filter? Rules might change and that would affect the deduplication that is not considering those commits.